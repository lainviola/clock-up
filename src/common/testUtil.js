// @flow
import React from 'react'

export function sel(id: string): string {
  return `[data-test="${id}"]`
}
