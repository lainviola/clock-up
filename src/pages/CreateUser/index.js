// @flow
import React from 'react'
import { withRouter, Redirect } from 'react-router-dom'
import { compose, pure } from 'recompose'
import { graphql } from 'react-apollo'
import type { MutationFunc } from 'react-apollo'
import { CREATE_USER_MUTATION } from '../../graphql/mutation'
import { CLOCK_BOARD_QUERY } from '../../graphql/query'
import { AUTH0_ID_TOKEN } from '../../index'
import Loading from '../../elements/Loading'
import type { WithRouterProps } from '../../propTypes'
import type { CLOCK_BOARD_QUERY_TYPE } from '../../graphql/query'

type Props = {
  ...WithRouterProps,
  ...CLOCK_BOARD_QUERY_TYPE,
  CreateUserMutation: MutationFunc<*, *>
}

export class CreateUser extends React.Component<Props> {
  isNotExistUserInAuth0(): boolean {
    const { data } = this.props

    return !data.user || window.localStorage.getItem(AUTH0_ID_TOKEN) !== null
  }

  insertUserDataToAuth0(): void {
    const { CreateUserMutation } = this.props

    const variables = {
      idToken: window.localStorage.getItem(AUTH0_ID_TOKEN)
    }

    // $FlowIssue
    CreateUserMutation({ variables }).catch(e => {
      if (
        e.message === 'GraphQL error: User already exists with that information'
      )
        return

      alert('error occurred when createUser')
    })
  }

  render() {
    const { data } = this.props

    if (data.loading) {
      return <Loading enzyme-testid="Loading" />
    }

    if (this.isNotExistUserInAuth0()) {
      this.insertUserDataToAuth0()
    }

    return (
      // $FlowIssue
      <Redirect
        to={{
          pathname: '/'
        }}
        enzyme-testid="Redirect"
      />
    )
  }
}

export default compose(
  graphql(CREATE_USER_MUTATION, { name: 'CreateUserMutation' }),
  graphql(CLOCK_BOARD_QUERY, {
    // $FlowIssue
    options: { fetchPolicy: 'network-only', notifyOnNetworkStatusChange: true }
  }),
  // $FlowIssue @see https://github.com/flowtype/flow-typed/issues/1914
  withRouter,
  pure
)(CreateUser)
