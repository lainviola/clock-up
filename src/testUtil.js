import React from 'react'

export function sel(id) {
  return `[data-test="${id}"]`
}
