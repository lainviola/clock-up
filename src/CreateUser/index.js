// @flow
import React from 'react'
import { withRouter, Redirect } from 'react-router-dom'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'
import { AUTH0_ID_TOKEN } from '../common/const'
import Loading from '../common/components/Loading'

type Props = {
  data: Object,
  createUser: Function
}

export class CreateUser extends React.Component<Props> {
  isNotExistUserInAuth0(): boolean {
    const { data } = this.props

    return !data.user || window.localStorage.getItem(AUTH0_ID_TOKEN) !== null
  }

  InsertUserDataToAuth0(): void {
    const { createUser } = this.props

    const variables = {
      idToken: window.localStorage.getItem(AUTH0_ID_TOKEN)
    }

    createUser({ variables }).catch(e => {
      if (
        e.message === 'GraphQL error: User already exists with that information'
      )
        return

      alert('error occurred when createUser')
    })
  }

  render() {
    const { data } = this.props

    if (data.loading) {
      return <Loading />
    }

    if (this.isNotExistUserInAuth0()) {
      this.InsertUserDataToAuth0()
    }

    return (
      <Redirect
        to={{
          pathname: '/'
        }}
      />
    )
  }
}

const createUser = gql`
  mutation($idToken: String!) {
    createUser(authProvider: { auth0: { idToken: $idToken } }) {
      id
    }
  }
`

const userQuery = gql`
  query {
    user {
      id
    }
  }
`

export default graphql(createUser, { name: 'createUser' })(
  graphql(userQuery, {
    options: { fetchPolicy: 'network-only', notifyOnNetworkStatusChange: true }
  })(withRouter(CreateUser))
)
