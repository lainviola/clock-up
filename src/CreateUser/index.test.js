import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import CreateUser from './index'
import { apolloWarapper } from '../testUtil'

describe('Component: CreateUser.index', () => {
  it('snapshot', () => {
    const wrapper = shallow(apolloWarapper(CreateUser))
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
