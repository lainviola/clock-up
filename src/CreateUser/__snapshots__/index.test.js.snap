// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CreateUser.index snapshot 1`] = `
<QueryRecyclerProvider>
  <Apollo(Apollo(withRouter(CreateUser))) />
</QueryRecyclerProvider>
`;
